### DISCUSSION TITLE
What does "Use vector encodings to automatically select code" in the roadmap mean?

### DISCUSSION THREAD

**Discussion by ziontee113** [jellydn/CopilotChat.nvim#26] [Q&A] [answered]
Hi, I found this "Use vector encodings to automatically select code" line on the README but can't quite understand what that means. Can you please explain it to me @gptlang?

**Comment by gptlang** [comment, marked as answer]
It's what the Copilot Chat extension does in VSCode. It takes the file you're on and uses tree-sitter to get all functions & classes. It then submits your code to the embedding API

https://github.com/jellydn/CopilotChat.nvim/blob/c54e61fd99bf1e9fe1fb6ccadd01766087d46645/rplugin/python3/copilot.py#L136-L150

Once that's done, it finds the `n` most similar code snippets using cosine distance on the embedding and automatically inserts that into the prompt rather than having to manually place code snippets into the clipboard/unnamed register.

Like this:
https://github.com/acheong08/vectordb/blob/83ee790411cdeb875d1db2ee29b81bb69eb9b368/rank/rank.go#L29-L61

**Comment by ziontee113** [comment]
Couldn't we just use Neovim built-in Treesitter and LSP to do that?

**Reply by gptlang** [reply]
Yes.

**Comment by ziontee113** [comment]
I'm really not sure why embeddings are necessary. What's the use case to use embeddings anyway?

**Reply by gptlang** [reply]
There is a `@workspace` command that allows it to insert multiple code segments relevant automatically from across your git repository. It's quite helpful in finding code that handles something when you don't know where it is.
